src/main/java/dev/miku/r2dbc/mysql/util/FluxDiscardOnCancel.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2018-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.miku.r2dbc.mysql.util;
+
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.CoreSubscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Operators;
+import reactor.util.context.Context;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A decorating operator that replays signals from its source to a {@link Subscriber} and drains the
+ * source upon {@link Subscription#cancel() cancel} and drops data signals until termination.
+ * Draining data is required to complete a particular request/response window and clear the protocol
+ * state as client code expects to start a request/response conversation without any previous
+ * response state.
+ * <p>
+ * This is a slightly altered version of R2DBC SQL Server's implementation:
+ * https://github.com/r2dbc/r2dbc-mssql
+ */
+final class FluxDiscardOnCancel<T> extends Flux<T> {
+
+    private static final Logger logger = LoggerFactory.getLogger(FluxDiscardOnCancel.class);
+
+    private final Flux<? extends T> source;
+
+    FluxDiscardOnCancel(Flux<? extends T> source) {
+        this.source = source;
+    }
+
+    @Override
+    public void subscribe(CoreSubscriber<? super T> actual) {
+        this.source.subscribe(new Subscriber<>(actual));
+    }
+
+    static final class Subscriber<T> extends AtomicBoolean implements CoreSubscriber<T>, Subscription {
+
+        final CoreSubscriber<T> actual;
+
+        final Context ctx;
+
+        Subscription s;
+
+        Subscriber(CoreSubscriber<T> actual) {
+            this.actual = actual;
+            this.ctx = actual.currentContext();
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            if (Operators.validate(this.s, s)) {
+                this.s = s;
+                this.actual.onSubscribe(this);
+            }
+        }
+
+        @Override
+        public void onNext(T t) {
+            if (get()) {
+                Operators.onDiscard(t, this.ctx);
+            } else {
+                this.actual.onNext(t);
+            }
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            this.actual.onError(t);
```

Comment:
I have a doubt: if `onError` is called by source, which means this is not user error (least not be `actual` subscriber error) and it should be source error, isn't it?

- If it is `actual` subscriber error, `onCancel` will be called by `actual` subscriber instead of `onError`.
- If it is `source` error, `onError` will be called by `source` and we can not receive anything, even we call `s.request(Long.MAX_VALUE)`.

Reply:
Not sure I follow. The perspective is: When a subscriber calls `cancel()` it no longer listens and propagates terminal signals. Therefore, errors would be silently dropped.

Reply:
Ah, that's clear. I misunderstood this.